library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt

```diff
@@ -194,40 +177,83 @@ internal class TransactionPayloadFragment :
+        } else {
+            adapter.resetHighlight()
+        }
         return true
     }
 
-    private class UiLoaderTask(val fragment: TransactionPayloadFragment) :
-        AsyncTask<Pair<Int, HttpTransaction>, Unit, UiPayload>() {
+    /**
+     * Async task responsible of loading in the background the content of the HTTP request/response.
+     */
+    class PayloadLoaderTask(private val fragment: TransactionPayloadFragment) :
+        AsyncTask<Pair<Int, HttpTransaction>, Unit, List<TransactionPayloadItem>>() {
+
+        override fun onPreExecute() {
+            val progressBar: ProgressBar? = fragment.view?.findViewById(R.id.progress_loading_transaction)
+            val recyclerView: RecyclerView? = fragment.view?.findViewById(R.id.transaction_content)
+            progressBar?.visibility = View.VISIBLE
+            recyclerView?.visibility = View.INVISIBLE
+        }
 
-        override fun doInBackground(vararg params: Pair<Int, HttpTransaction>): UiPayload {
+        @Suppress("ComplexMethod")
+        override fun doInBackground(vararg params: Pair<Int, HttpTransaction>): List<TransactionPayloadItem> {
             val (type, transaction) = params[0]
-            return if (type == TYPE_REQUEST) {
-                UiPayload(
-                    transaction.getRequestHeadersString(true),
-                    transaction.getFormattedRequestBody(),
-                    transaction.isRequestBodyPlainText
-                )
+            val result = mutableListOf<TransactionPayloadItem>()
+
+            val headersString: String
+            val isBodyPlainText: Boolean
+            val bodyString: String
+
+            if (type == TYPE_REQUEST) {
+                headersString = transaction.getRequestHeadersString(true)
+                isBodyPlainText = transaction.isRequestBodyPlainText
+                bodyString = transaction.getFormattedRequestBody()
             } else {
-                UiPayload(
-                    transaction.getResponseHeadersString(true),
-                    transaction.getFormattedResponseBody(),
-                    transaction.isResponseBodyPlainText,
-                    transaction.responseImageBitmap
+                headersString = transaction.getResponseHeadersString(true)
+                isBodyPlainText = transaction.isResponseBodyPlainText
+                bodyString = transaction.getFormattedResponseBody()
+            }
+
+            if (headersString.isNotBlank()) {
+                result.add(
+                    TransactionPayloadItem.HeaderItem(
+                        HtmlCompat.fromHtml(headersString, HtmlCompat.FROM_HTML_MODE_LEGACY)
+                    )
                 )
             }
+
+            // The body could either be an image, binary encoded or plain text.
+            if (type == TYPE_RESPONSE && transaction.responseImageBitmap != null) {
+                transaction.responseImageBitmap?.let {
```

Comment:
Isn't `responseImageBitmap` already checked for `null`?

Reply:
`responseImageBitmap` is actually a property with a custom getter. Resolved storing the value in an immutable variable.